packages/core/src/__tests__/state-manager.test.ts

```diff
@@ -71,75 +89,151 @@ describe('anchor', () => {
-    fakeHandleTip.mockImplementationOnce(realHandleTip)
-    fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+    const bound = fakeHandleTip.bind(stateManager)
+    ;(stateManager as any)._handleTip = bound
+
+    // Mock a throw as the first call
     fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+
+    // Mock the result of the original implementation as the second call - this one should return
+    // and stop the retrying mechanism
     fakeHandleTip.mockImplementationOnce(realHandleTip)
 
-    await new Promise<void>((resolve) => {
-      ceramic.repository.stateManager.anchor(stream$).add(resolve)
-    })
-    expect(stream$.value.anchorStatus).toEqual(AnchorStatus.ANCHORED)
-    expect(fakeHandleTip).toHaveBeenCalledTimes(4)
+    // wait for 3 secs to let the retry mechanism work
+    await TestUtils.delay(3000)
+
+    // Check that fakeHandleTip was called only three times
+    expect(fakeHandleTip).toHaveBeenCalledTimes(2)
   })
 
-  test('anchor too many retries', async () => {
+  test(`_handleTip is retried up to three times, if it doesn't return`, async () => {
     const stateManager = ceramic.repository.stateManager
     const stream = await TileDocument.create(ceramic, INITIAL_CONTENT, null, { anchor: false })
     const stream$ = await ceramic.repository.load(stream.id, {})
 
+    await ceramic.repository.stateManager.anchor(stream$)
+    expect(stream$.value.anchorStatus).toEqual(AnchorStatus.PENDING)
+    await waitForAnchorStatus(stream$, AnchorStatus.ANCHORED)
+
     const fakeHandleTip = jest.fn()
+    fakeHandleTip.bind(stateManager)
     ;(stateManager as any)._handleTip = fakeHandleTip
+
+    // Mock a bunch of throws as consecutive calls to fakeHandleTip
     fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
     fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
     fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
-    fakeHandleTip.mockImplementationOnce(realHandleTip)
+    fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+    fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+    fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+
+    // wait for 3 secs to let the retry mechanism work
+    await TestUtils.delay(3000)
+
+    // Check that fakeHandleTip was called only three times
+    expect(fakeHandleTip).toHaveBeenCalledTimes(3)
+  })
+
+  test('anchor request is stored when created and deleted when anchored', async () => {
+    const stream = await TileDocument.create(ceramic, INITIAL_CONTENT, null, { anchor: false })
+    const stream$ = await ceramic.repository.load(stream.id, {})
+
+    // @ts-ignore anchorRequestStore is private
+    const anchorRequestStore = ceramic.repository.stateManager.anchorRequestStore
+
+    expect(await anchorRequestStore.load(stream.id)).toBeNull()
+    await ceramic.repository.stateManager.anchor(stream$)
```

Comment:
shouldn't this line be after we check for AnchorStatus of PENDING and that the request store has the pending request (ie right before we call `waitForAnchorStatus`)?  Otherwise isn't there a race condition if the anchor happens too quickly?

Reply:
You mean that
```
 await ceramic.repository.stateManager.anchor(stream$)
 ```
 
 should be after
 ```
 expect(stream$.value.anchorStatus).toEqual(AnchorStatus.PENDING)
 ```
 ?
 
 The status of the anchor won't be `PENDING`, before we call `anchor(...)`.
 
 If I misunderstand you, can you propose code change, please?

Reply:
> ah, so we're relying on the call to anchor() to send the stream through both PENDING and ANCHORED statuses? 

No, the only thing I wrote was that `stream$.value.anchorStatus` will not be `AnchorStatus.PENDING`, before the call to `await ceramic.repository.stateManager.anchor(stream$)`

I think I really don't understand the point you originally made in this comment thread, @stbrody . Can you please create a commit suggestion here with the particular contents of this test that you have in mind :) ?

Reply:
I think you could fix this by using a Ceramic node with `anchorOnRequest: false` set for the InMemoryAnchorService, like you do in some of the other tests in this file.  That way you're guaranteed that it won't move past PENDING into ANCHORED until you explicitly trigger the anchor against the InMemoryAnchorService